### DISCUSSION TITLE
Feature Importance and early stopping for ML models such as XGBoost

### DISCUSSION THREAD

**Discussion by alexanderlange53** [alan-turing-institute/sktime#3210] [Q&A]
Hello everyone,

thanks for creating this great library, it is really helpful!

I have two questions:

1. It would be very useful to allow easy checking of features importance.  Currently, it looks like the feature names get lost during the transformation process, or am I missing something?

2. Is it possible to specify a validation set for early stopping when fitting a model like XGBoost?

**Comment by fkiraly** [comment]
> It would be very useful to allow easy checking of features importance. Currently, it looks like the feature names get lost during the transformation process, or am I missing something?

That is a known issue.
We are currently re-working the reducers (@danbartl and myself), and the new version will be `pandas` based and retain feature names.

You can find one of the prototypes here, feedback is appreciated:

https://github.com/alan-turing-institute/sktime/pull/2833

Thinking of releasing this as experimental for user feedback, but it is not the final thing.
Wrapping `xgboost` in this should produce the feature importances with variable names.

@danbartl might be able to point to other work on the reducers.

> Is it possible to specify a validation set for early stopping when fitting a model like XGBoost?

You mean, via `eval_set` as passed to `fit`?

I think not, as `xgboost` is not `sklearn` compliant here, it has extra arguments in `fit` that are not expected in an `sklearn` interface, and some of them are also not exposed by `xgboost`'s `sklearn` compatible interface (such as `eval_set`).

This is a known interface deficiency of `xgboost` that doesn't have to do with `sktime` as far as I know.

If you want that, my recommendation is to open an issue or contribute to `xgboost` directly, specifically to the `sktime` interface.

Here's how I would address it: I would add an `__init__` argument `eval_sets`, which is not a list of data containers, but a list of `pandas` index, or list of iterable of integer, each element of which specifies `loc` (`pandas` index) or `iloc` (iterable of integer) sub-sets, as a reference to `X` and `y` passed to `fit`, to be used in validation. Could also be a list of float, in which case the last x percent would be used in validation.

This could be paired by another arg `train_sets`, if you want to have disjoint train/validation sets for early stopping.

**Reply by alexanderlange53** [reply]
Thank you very much!
